Luabind "you are trying to register a class twice" Error in Extractor

root@/osrm_log #  ./osrm-extract -p profile.lua gcc-states-latest.osm.pbf 
[info] Using script profile.lua
[info] Input file: gcc-states-latest.osm.pbf
[info] Profile: profile.lua
[info] Threads: 24
[STXXL-MSG] STXXL v1.4.99 (prerelease/Release) (git 1babe452214b4613a2a488d80073f4185c05a0b3) + gnu parallel(20150623)
[STXXL-MSG] Disk '/tmp/stxx' is allocated, space: 50000 MiB, I/O implementation: syscall queue=0 devid=0
[info] Parsing in progress..
osrm-extract: src/class_registry.cpp:152: void luabind::detail::class_registry::add_class(const luabind::type_id&, luabind::detail::class_rep*): Assertion `(m_classes.find(info) == m_classes.end()) && "you are trying to register a class twice"' failed.

Nice work @daniel-j-h! I confirmed that your fix works locally. To test this I built luabind in Debug by doing:

```
diff --git a/third_party/mason/scripts/luabind_lua524/e414c57bcb687bb3091b7c55bbff6947f052e46b/script.sh b/third_party/mason/scripts/luabind_lua524/e414c57bcb687bb3091b7c55bbff6947f052e46b/script.sh
index 697d726..7030aef 100755
--- a/third_party/mason/scripts/luabind_lua524/e414c57bcb687bb3091b7c55bbff6947f052e46b/script.sh
+++ b/third_party/mason/scripts/luabind_lua524/e414c57bcb687bb3091b7c55bbff6947f052e46b/script.sh
@@ -35,7 +35,7 @@ function mason_compile {
       -DLUA_LIBRARIES=${MASON_LUA}/lib \
       -DLUA_INCLUDE_DIR=${MASON_LUA}/include \
       -DBOOST_INCLUDEDIR=${MASON_BOOST_HEADERS}/include \
-      -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_BUILD_TYPE=Debug \
       -DBUILD_TESTING=OFF
     make -j${MASON_CONCURRENCY} VERBOSE=1
     make install
```

And then:

```
./third_party/mason/mason build luabind_lua524 e414c57bcb687bb3091b7c55bbff6947f052e46b
```

And then build osrm-backend against this luabind with `-DENABLE_MASON=ON`. That replicated the problem again with latest master:

```
../../scripts/timer.sh "osrm-extract" /Users/dane/projects/osrm-backend/build-osrm/osrm-extract monaco.osm.pbf -p ../../profiles/car.lua
[info] Using script ../../profiles/car.lua
[info] Input file: monaco.osm.pbf
[info] Profile: car.lua
[info] Threads: 4
[STXXL-MSG] STXXL v1.4.1 (prerelease/Release)
[STXXL-ERRMSG] Warning: no config file found.
[STXXL-ERRMSG] Using default disk configuration.
[STXXL-MSG] Disk '/var/tmp/stxxl' is allocated, space: 1000 MiB, I/O implementation: syscall autogrow delete_on_exit queue=0 devid=0
[info] Parsing in progress..
Assertion failed: ((m_classes.find(info) == m_classes.end()) && "you are trying to register a class twice"), function add_class, file /Users/dane/projects/osrm-backend/mason_packages/.build/luabind-e414c57bcb687bb3091b7c55bbff6947f052e46b/src/class_registry.cpp, line 126.
../../scripts/timer.sh: line 11: 19019 Abort trap: 6           /bin/bash -c "$CMD"
make: *** [monaco.osrm] Error 134
```

But patching master with https://patch-diff.githubusercontent.com/raw/Project-OSRM/osrm-backend/pull/3346.diff fixed the problem upon rebuild.

--- a/src/extractor/scripting_environment_lua.cpp
+++ b/src/extractor/scripting_environment_lua.cpp
@@ -35,9 +35,12 @@ template <class T>
 auto get_value_by_key(T const &object, const char *key) -> decltype(object.get_value_by_key(key))
 {
     auto v = object.get_value_by_key(key);
-    if (v && *v) {             // non-empty string?
+    if (v && *v)
+    { // non-empty string?
         return v;
-    } else {
+    }
+    else
+    {
         return NULL;
     }
 }
@@ -82,6 +85,23 @@ void LuaScriptingEnvironment::InitContext(LuaScriptingContext &context)
 
     util::luaAddScriptFolderToLoadPath(context.state, file_name.c_str());
 
+    // In the following we have to register several char typedefs to Lua which we can't
+    // make proper enums (since enums can't be bit-packed but chars can).
+    //
+    // Luabind crashes hard (sometimes on some platforms) during runtime when registering
+    // the same type (e.g. our char typedefs) multiple times.
+    //
+    // We work around this by creating separate enum types for all typedefs here.
+    // This makes sure they have different typeid()s which Luabind dispatches upon.
+    //
+    // We still bind the typedef values to the enum since the enum is integral by default.
+    enum TravelModeEnumDummyForLuabind
+    {
+    };
+    enum RoadPriorityClassEnumDummyForLuabind
+    {
+    };
+
     // Add our function to the state's global scope
     luabind::module(context.state)
         [luabind::def("durationIsValid", durationIsValid),
@@ -89,7 +109,7 @@ void LuaScriptingEnvironment::InitContext(LuaScriptingContext &context)
          luabind::def("trimLaneString", trimLaneString),
          luabind::def("applyAccessTokens", applyAccessTokens),
          luabind::def("canonicalizeStringList", canonicalizeStringList),
-         luabind::class_<TravelMode>("mode").enum_(
+         luabind::class_<TravelModeEnumDummyForLuabind>("mode").enum_(
              "enums")[luabind::value("inaccessible", TRAVEL_MODE_INACCESSIBLE),
                       luabind::value("driving", TRAVEL_MODE_DRIVING),
                       luabind::value("cycling", TRAVEL_MODE_CYCLING),
@@ -103,7 +123,7 @@ void LuaScriptingEnvironment::InitContext(LuaScriptingContext &context)
                       luabind::value("river_down", TRAVEL_MODE_RIVER_DOWN),
                       luabind::value("route", TRAVEL_MODE_ROUTE)],
 
-         luabind::class_<extractor::guidance::RoadPriorityClass::Enum>("road_priority_class")
+         luabind::class_<RoadPriorityClassEnumDummyForLuabind>("road_priority_class")
              .enum_("enums")
                  [luabind::value("motorway", extractor::guidance::RoadPriorityClass::MOTORWAY),
                   luabind::value("trunk", extractor::guidance::RoadPriorityClass::TRUNK),
@@ -188,10 +208,17 @@ void LuaScriptingEnvironment::InitContext(LuaScriptingContext &context)
              .def_readwrite("backward_speed", &ExtractionWay::backward_speed)
              .property("name", &ExtractionWay::GetName, &ExtractionWay::SetName)
              .property("ref", &ExtractionWay::GetRef, &ExtractionWay::SetRef)
-             .property("pronunciation", &ExtractionWay::GetPronunciation, &ExtractionWay::SetPronunciation)
-             .property("destinations", &ExtractionWay::GetDestinations, &ExtractionWay::SetDestinations)
-             .property("turn_lanes_forward", &ExtractionWay::GetTurnLanesForward, &ExtractionWay::SetTurnLanesForward)
-             .property("turn_lanes_backward", &ExtractionWay::GetTurnLanesBackward, &ExtractionWay::SetTurnLanesBackward)
+             .property("pronunciation",
+                       &ExtractionWay::GetPronunciation,
+                       &ExtractionWay::SetPronunciation)
+             .property(
+                 "destinations", &ExtractionWay::GetDestinations, &ExtractionWay::SetDestinations)
+             .property("turn_lanes_forward",
+                       &ExtractionWay::GetTurnLanesForward,
+                       &ExtractionWay::SetTurnLanesForward)
+             .property("turn_lanes_backward",
+                       &ExtractionWay::GetTurnLanesBackward,
+                       &ExtractionWay::SetTurnLanesBackward)
              .def_readwrite("roundabout", &ExtractionWay::roundabout)
              .def_readwrite("is_access_restricted", &ExtractionWay::is_access_restricted)
              .def_readwrite("is_startpoint", &ExtractionWay::is_startpoint)

